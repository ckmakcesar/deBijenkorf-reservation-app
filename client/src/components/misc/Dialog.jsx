import React, { useRef, useEffect } from 'react';
import clsx from 'clsx';
import PropTypes from 'prop-types';

import Header from './Header';
import Footer from './Footer';
import Button from './Button';
import { isEscKeyDown } from '../../utils/utils';

import styles from '../../styles/misc/Dialog.module.css';

const Dialog = ({
  isOpen,
  headerTitle,
  confirmButtonText,
  confirmButtonTone,
  onClose,
  onConfirm,
  children,
}) => {
  if (!isOpen) return null;

  const handleConfirm = (payload) => {
    onConfirm(payload);
    onClose();
  };

  // set focus to the overlay, and quit if ESCAPE is pressed
  const focusDiv = useRef();
  useEffect(()=>{
    if (isOpen && focusDiv.current) {
      focusDiv.current.focus();
    }
  },[isOpen]);
  const handleKeyDown = (e) => {
    if (isEscKeyDown(e)) {
      onClose();
    }
  };

  return (
    <div
      className={styles.dialogBackground}
      onClick={onClose} // clicking the opaque background will close the dialog
      ref={focusDiv}
      tabIndex={0}
      onKeyDown={handleKeyDown}
    >
      <div
        className={clsx(styles.dialogBody, isOpen && styles.openedDialog)}
        onClick={(e) => e.stopPropagation()} // pressing on dialog body should not close the dialog
      >
        <Header text={headerTitle} />

        {/* drawer content -> flex & overflow scrol */}
        <div className='flex-content'>
          {children}
        </div>

        <Footer reducedPadding>
          <div className={styles.dialogButtons}>
            <Button
              className='flex-grow'
              text={confirmButtonText}
              onClick={handleConfirm}
              tone={confirmButtonTone}
            />

            <Button
              className='flex-grow'
              onClick={() => { onClose() }}
              text='Cancel'
              secondary
            />
          </div>
        </Footer>
      </div>

    </div>
  );
}

// *********************************************************************************************************************
// PROPS
// *********************************************************************************************************************
Dialog.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  headerTitle: PropTypes.string.isRequired,
  confirmButtonText: PropTypes.string,
  confirmButtonTone: PropTypes.string,
  onClose: PropTypes.func.isRequired,
  onConfirm: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
};

Dialog.defaultProps = {
  confirmButtonText: 'Confirm',
  confirmButtonTone: '',
};

export default Dialog;