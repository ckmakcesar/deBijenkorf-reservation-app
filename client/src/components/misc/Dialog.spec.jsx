import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Dialog from './Dialog';

describe('component:Dialog', () => {
  it('Should render Dialog with every optional props absent', () => {
    const wrapper = shallow(
      <Dialog
        isOpen={true}
        headerTitle='Top of Page'
        onClose={jest.fn()}
        onConfirm={jest.fn()}
      >
        {/* lint recommends NOT to nest children as prop */}
        <h5>hello world</h5>
      </Dialog>
      ,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('Should render Dialog with every optional props present', () => {
    const wrapper = shallow(
      <Dialog
        isOpen={true}
        headerTitle='Top of Page'
        confirmButtonText='Vamos Amigo'
        confirmButtonTone='green'
        onClose={jest.fn()}
        onConfirm={jest.fn()}
      >
        <h5>hello world</h5>
      </Dialog>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
