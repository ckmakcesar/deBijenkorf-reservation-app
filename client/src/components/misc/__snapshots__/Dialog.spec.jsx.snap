// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:Dialog Should render Dialog with every optional props absent 1`] = `
<div
  className="dialogBackground"
  onClick={[MockFunction]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="dialogBody openedDialog"
    onClick={[Function]}
  >
    <Header
      id=""
      smaller={true}
      text="Top of Page"
    />
    <div
      className="flex-content"
    >
      <h5>
        hello world
      </h5>
    </div>
    <Footer
      id=""
      reducedPadding={true}
    >
      <div
        className="dialogButtons"
      >
        <Button
          boxIconClassName=""
          className="flex-grow"
          disabled={false}
          formId=""
          id=""
          isSubmit={false}
          onClick={[Function]}
          text="Confirm"
          tone=""
        />
        <Button
          boxIconClassName=""
          className="flex-grow"
          disabled={false}
          formId=""
          id=""
          isSubmit={false}
          onClick={[Function]}
          text="Cancel"
          tone="secondary"
        />
      </div>
    </Footer>
  </div>
</div>
`;

exports[`component:Dialog Should render Dialog with every optional props present 1`] = `
<div
  className="dialogBackground"
  onClick={[MockFunction]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="dialogBody openedDialog"
    onClick={[Function]}
  >
    <Header
      id=""
      smaller={true}
      text="Top of Page"
    />
    <div
      className="flex-content"
    >
      <h5>
        hello world
      </h5>
    </div>
    <Footer
      id=""
      reducedPadding={true}
    >
      <div
        className="dialogButtons"
      >
        <Button
          boxIconClassName=""
          className="flex-grow"
          disabled={false}
          formId=""
          id=""
          isSubmit={false}
          onClick={[Function]}
          text="Vamos Amigo"
          tone="green"
        />
        <Button
          boxIconClassName=""
          className="flex-grow"
          disabled={false}
          formId=""
          id=""
          isSubmit={false}
          onClick={[Function]}
          text="Cancel"
          tone="secondary"
        />
      </div>
    </Footer>
  </div>
</div>
`;
